DB2: Fetch first row only gives exception (also with UR problem)

### Actual Behavior
This code gives ParseException:

    	String str = 
    			"select distinct T1.A  " 
    			+" from  T2 join  T1  on (T2.B = T1.C)  "
    		    + " fetch first row only with UR";
    			//  + "fetch first 3 rows only with UR";
    	Statement stmt1 = CCJSqlParserUtil.parse(str);

Exception:
	
	Exception in thread "main" net.sf.jsqlparser.JSQLParserException
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:55)
	at csn.mier.test.sqlparser.App.main(App.java:56)
	Caused by: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "row" "ROW"
	at line 1, column 74.
	
	Was expecting one of:
	
	"?"
	<S_LONG>
	
	at net.sf.jsqlparser.parser.CCJSqlParser.generateParseException(CCJSqlParser.java:18047)
	at net.sf.jsqlparser.parser.CCJSqlParser.jj_consume_token(CCJSqlParser.java:17900)


Note that "FETCH FIRST 1 ROW ONLY" works, but then adding "WITH UR" gives another exception.


	net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "with" "WITH"
	    at line 1, column 85.
	
	Was expecting one of:
	
	    ";"
	    "EXCEPT"
	    "FOR"
	    "INTERSECT"
	    "MINUS"
	    "ORDER"
	    "UNION"
	    <EOF>
	
		at net.sf.jsqlparser.parser.CCJSqlParser.generateParseException(CCJSqlParser.java:18047)
		at net.sf.jsqlparser.parser.CCJSqlParser.jj_consume_token(CCJSqlParser.java:17900)
		at net.sf.jsqlparser.parser.CCJSqlParser.Statement(CCJSqlParser.java:91)
		at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:53)


### Expected Behavior

Not get an exception

### Steps to Reproduce the Problem

Run the code above

### Specifications

  - Version: 1.2
  - Platform: Windows
  - Subsystem: NA

--- a/src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java
@@ -54,6 +54,7 @@ public class PlainSelect extends ASTNodeAccessImpl implements SelectBody {
     private KSQLWindow ksqlWindow = null;
     private boolean noWait = false;
     private boolean emitChanges = false;
+    private WithIsolation withIsolation;
 
     public boolean isUseBrackets() {
         return useBrackets;
@@ -330,6 +331,15 @@ public class PlainSelect extends ASTNodeAccessImpl implements SelectBody {
         return emitChanges;
     }
 
+    
+    public WithIsolation getWithIsolation() {
+        return withIsolation;
+    }
+
+    public void setWithIsolation(WithIsolation withIsolation) {
+        this.withIsolation = withIsolation;
+    }
+
     @Override
     @SuppressWarnings({"PMD.CyclomaticComplexity" , "PMD.ExcessiveMethodLength", "PMD.NPathComplexity"})
     public String toString() {
@@ -421,6 +431,10 @@ public class PlainSelect extends ASTNodeAccessImpl implements SelectBody {
             if (fetch != null) {
                 sql.append(fetch);
             }
+
+            if (withIsolation != null) {
+                sql.append(withIsolation);
+            }
             if (isForUpdate()) {
                 sql.append(" FOR UPDATE");
 
@@ -455,6 +469,9 @@ public class PlainSelect extends ASTNodeAccessImpl implements SelectBody {
             if (fetch != null) {
                 sql.append(fetch);
             }
+            if (withIsolation != null) {
+                sql.append(withIsolation);
+            }            
         }
         if (forXmlPath != null) {
             sql.append(" FOR XML PATH(").append(forXmlPath).append(")");

--- a/src/main/java/net/sf/jsqlparser/statement/select/SetOperationList.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/SetOperationList.java
@@ -24,6 +24,7 @@ public class SetOperationList implements SelectBody {
     private Limit limit;
     private Offset offset;
     private Fetch fetch;
+    private WithIsolation withIsolation;
 
     @Override
     public void accept(SelectVisitor selectVisitor) {
@@ -96,7 +97,16 @@ public class SetOperationList implements SelectBody {
         this.fetch = fetch;
     }
 
+    public Fetch getWithIsolation() {
+        return fetch;
+    }
+
+    public void setWithIsolation(WithIsolation withIsolation) {
+        this.withIsolation = withIsolation;
+    }
+
     @Override
+    @SuppressWarnings({"PMD.CyclomaticComplexity"})
     public String toString() {
         StringBuilder buffer = new StringBuilder();
 
@@ -122,6 +132,9 @@ public class SetOperationList implements SelectBody {
         }
         if (fetch != null) {
             buffer.append(fetch.toString());
+        }
+        if (withIsolation != null) {
+            buffer.append(withIsolation.toString());
         }
         return buffer.toString();
     }

--- a/src/main/java/net/sf/jsqlparser/statement/select/WithIsolation.java
+++ b/src/main/java/net/sf/jsqlparser/statement/select/WithIsolation.java
@@ -1,0 +1,28 @@
+/*-
+ * #%L
+ * JSQLParser library
+ * %%
+ * Copyright (C) 2004 - 2019 JSQLParser
+ * %%
+ * Dual licensed under GNU LGPL 2.1 or Apache License 2.0
+ * #L%
+ */
+package net.sf.jsqlparser.statement.select;
+
+
+public class WithIsolation {
+
+    private String isolation = "UR";
+
+    public String getIsolation() {
+        return this.isolation;
+    }
+    public void setIsolation(String s) {
+        this.isolation = s;
+    }
+
+    @Override
+    public String toString() {
+        return " WITH " + this.isolation;
+    }
+}

--- a/src/main/java/net/sf/jsqlparser/util/deparser/SelectDeParser.java
+++ b/src/main/java/net/sf/jsqlparser/util/deparser/SelectDeParser.java
@@ -199,6 +199,9 @@ public class SelectDeParser extends AbstractDeParser<PlainSelect>
         if (plainSelect.getFetch() != null) {
             deparseFetch(plainSelect.getFetch());
         }
+        if (plainSelect.getWithIsolation() != null) {
+            buffer.append(plainSelect.getWithIsolation().toString());
+        }
         if (plainSelect.isForUpdate()) {
             buffer.append(" FOR UPDATE");
             if (plainSelect.getForUpdateTable() != null) {
@@ -479,6 +482,9 @@ public class SelectDeParser extends AbstractDeParser<PlainSelect>
         }
         if (list.getFetch() != null) {
             deparseFetch(list.getFetch());
+        }
+        if (list.getWithIsolation() != null) {
+            buffer.append(list.getWithIsolation().toString());
         }
     }
 

--- a/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt
+++ b/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt
@@ -212,6 +212,7 @@ TOKEN: /* SQL Keywords. prefixed with K_ to avoid name clashes */
 |   <K_EXTENDED:"EXTENDED">
 |   <K_EXTRACT:"EXTRACT">
 |   <K_FETCH:"FETCH">
+|   <K_ISOLATION:("UR" | "RS" | "RR" | "CS")>
 |   <K_FILTER: "FILTER">
 |   <K_FIRST: "FIRST">
 |   <K_FALSE: "FALSE">
@@ -1808,6 +1809,7 @@ PlainSelect PlainSelect() #PlainSelect:
     Limit limit = null;
     Offset offset = null;
     Fetch fetch = null;
+    WithIsolation withIsolation = null;
     OptimizeFor optimize = null;
     Top top = null;
         Skip skip = null;
@@ -1875,7 +1877,7 @@ PlainSelect PlainSelect() #PlainSelect:
     [LOOKAHEAD(<K_OFFSET>) offset = Offset() { plainSelect.setOffset(offset);    } ]
 	[LOOKAHEAD(<K_LIMIT>, { limit==null }) limit = LimitWithOffset() { plainSelect.setLimit(limit);    } ]
     [LOOKAHEAD(<K_FETCH>) fetch = Fetch() { plainSelect.setFetch(fetch);    } ]
-
+    [LOOKAHEAD(<K_WITH> <K_ISOLATION>) withIsolation = WithIsolation() { plainSelect.setWithIsolation(withIsolation);    } ]
     [LOOKAHEAD(2) <K_FOR> <K_UPDATE> { plainSelect.setForUpdate(true); }
         [ <K_OF> updateTable = Table() { plainSelect.setForUpdateTable(updateTable); } ]
         [ LOOKAHEAD(<K_WAIT>) wait = Wait() { plainSelect.setWait(wait); } ]
@@ -1901,6 +1903,7 @@ SelectBody SetOperationList() #SetOperationList: {
     Limit limit = null;
     Offset offset = null;
     Fetch fetch = null;
+    WithIsolation withIsolation = null;
     SelectBody select = null;
     List<SelectBody> selects = new ArrayList<SelectBody>();
     List<SetOperation> operations = new ArrayList<SetOperation>();
@@ -1925,6 +1928,7 @@ SelectBody SetOperationList() #SetOperationList: {
     [LOOKAHEAD(<K_LIMIT>) limit=LimitWithOffset() {list.setLimit(limit);} ]
     [LOOKAHEAD(<K_OFFSET>) offset = Offset() { list.setOffset(offset);} ]
     [LOOKAHEAD(<K_FETCH>) fetch = Fetch() { list.setFetch(fetch);} ]
+    [LOOKAHEAD(<K_WITH> <K_ISOLATION>) withIsolation = WithIsolation() { list.setWithIsolation(withIsolation);} ]
 
     {
         if (selects.size()==1 && selects.get(0) instanceof PlainSelect && orderByElements==null) {
@@ -1956,6 +1960,7 @@ SelectBody SetOperationListWithoutIntialSelect(FromItem fromItem) #SetOperationL
     Limit limit = null;
     Offset offset = null;
     Fetch fetch = null;
+    WithIsolation withIsolation = null;
     SelectBody select;
     List<SelectBody> selects = new ArrayList<SelectBody>();
     List<SetOperation> operations = new ArrayList<SetOperation>();
@@ -2832,6 +2837,24 @@ Fetch Fetch():
         )
     {
         return fetch;
+    }
+}
+
+WithIsolation WithIsolation():
+{
+    WithIsolation withIsolation = new WithIsolation();
+    Token token = null;
+    JdbcParameter jdbc;
+}
+{
+    (
+            //with (ur | cs | rs | rr)
+             <K_WITH>
+             token=<K_ISOLATION> { withIsolation.setIsolation(token.image); }
+
+        )
+    {
+        return withIsolation;
     }
 }
 

--- a/src/test/java/net/sf/jsqlparser/statement/select/SelectTest.java
+++ b/src/test/java/net/sf/jsqlparser/statement/select/SelectTest.java
@@ -4980,4 +4980,19 @@ public class SelectTest {
     public void testIgnoreNullsForWindowFunctionsIssue1429() throws JSQLParserException {
         assertSqlCanBeParsedAndDeparsed("SELECT lag(mydata) IGNORE NULLS OVER (ORDER BY sortorder) AS previous_status FROM mytable");
     }
+
+    @Test
+    public void testWithIsolation() throws JSQLParserException {
+        String statement = "SELECT * FROM mytable WHERE mytable.col = 9 WITH ur";
+        Select select = (Select) parserManager.parse(new StringReader(statement));
+        String isolation = ((PlainSelect) select.getSelectBody()).getWithIsolation().getIsolation();
+        assertEquals("ur", isolation);
+        assertSqlCanBeParsedAndDeparsed(statement);
+
+        statement = "SELECT * FROM mytable WHERE mytable.col = 9 WITH Cs";
+        select = (Select) parserManager.parse(new StringReader(statement));
+        isolation = ((PlainSelect) select.getSelectBody()).getWithIsolation().getIsolation();
+        assertEquals("Cs", isolation);
+        assertSqlCanBeParsedAndDeparsed(statement);
+    }
 }

